rfd/0155-scoped-webauthn-credentials.md

```diff
@@ -0,0 +1,277 @@
++  // Used for per-session MFA and moderated session presence checks.
++  SCOPE_SESSION = 5;
++  // Headless login approval.
++  SCOPE_HEADLESS = 6;
++  // Used for various administrative actions, such as adding, updating, or
++  // deleting administrative resources (users, roles, etc.).
++  //
++  // Note: this scope should not be used for new MFA capabilities that have
++  // more precise scope. Instead, new scopes should be added. This scope may
++  // also be split into multiple smaller scopes in the future.
++  SCOPE_ADMIN_ACTION = 7;
++}
+
+// authservice.proto
+
+message CreateAuthenticateChallengeRequest {
+  oneof Request {
+    UserCredentials UserCredentials = 1 [(gogoproto.jsontag) = "user_credentials,omitempty"];
+    string RecoveryStartTokenID = 2 [(gogoproto.jsontag) = "recovery_start_token_id,omitempty"];
+    ContextUser ContextUser = 3 [(gogoproto.jsontag) = "context_user,omitempty"];
+    Passwordless Passwordless = 4 [(gogoproto.jsontag) = "passwordless,omitempty"];
+  }
+  IsMFARequiredRequest MFARequiredCheck = 5 [(gogoproto.jsontag) = "mfa_required_check,omitempty"];
+
++  // Scope is a authorization scope for this MFA challenge.
++  // Required. Only applies to webauthn challenges.
++  webauthn.ChallengeScope Scope = 6 [(gogoproto.jsontag) = "scope,omitempty"];
++  // AllowReuse means webauthn credentials resolved from this challenge can be
++  // reused for a short span of time before the challenge expires.
++  //
++  // Reuse is only permitted for specific actions by the discretion of the server.
++  // See the server implementation for details.
++  bool AllowReuse = 7 [(gogoproto.jsontag) = "allow_reuse,omitempty"];
+}
+```
+
+### Client changes
+
+Clients will be expected to provide a `Scope` when requesting an MFA challenge
+through `rpc CreateAuthenticateChallenge`. For specific login and device
+management endpoints, the scope will be automatically set on the server side.
+
+Clients can optionally provide `AllowReuse=true` if the client wants to reuse
+the resulting webauthn credentials for multiple requests. However, the server
+will only permit reuse for a select number of actions. See
+[the server implementation details](#reuse).
+
+### Server changes
+
+#### Scope
+
+When verifying a webauthn credential against the user's stored webauthn
+challenge, the Auth Server will check that the stored scope matches the scope
+that the Auth server is verifying for. For example, if the Auth server is
+verifying a webauthn credential for scope "login", but webauthn challenge
+stored for the user has scope "headless", the verification will fail.
+
+#### Reuse
+
+If an MFA action does not allow reuse, the Auth Server will validate that the
+webauthn challenge has `AllowReuse=false`.
+
+Additionally, challenges with `AllowReuse=true` will not be deleted immediately,
+instead letting them expire in the backend after 5 minutes.
+
+Initially, reuse will only be permitted for the following admin action RPCs:
```

Comment:
We're going to have to handle it on a case by case basis, but hopefully it will be clear in the code with comments. I'll add a short section explaining our approach.